### PR TITLE
[SD-LIE] Common tools

### PR DESCRIPTION
**Pull request by norihiro-w**
As #1452 is quite big (>3k lines), I would like to ask code review piece by piece. This first piece is focusing on common tools for the process, located under ProcessLib/SmallDeformationWithLIE/Common. 

The directory includes the followings
- Fracture property class and its configuration
- H-matrix tools similar to B-matrix (H-matrix is based on shape vectors)
- Levelset function
- Mesh tools for getting information about fracture and rock matrix elements
- Other tools for computing a normal vector and a rotation matrix, and for computing physical coordinates from integration point coordinates

TODO
- [ ] replace `#pragma once`

### CODE REVIEW

File: ProcessLib/SmallDeformationWithLIE/Common/HMatrixUtils.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace ProcessLib
+{
+
+/// An implementation of H-Matrix policy using same matrix and vector types
+/// (fixed size or dynamic) as in the ShapeMatrixPolicyType.
+template <typename ShapeFunction, unsigned DisplacementDim>
+class HMatrixPolicyType
+{
+private:
```
Comment (wenqing) [inline ProcessLib/SmallDeformationWithLIE/Common/HMatrixUtils.h]: private section is better at the bottom.

File: ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    // which are passed to a DoF table
+    // first, collect matrix elements
+    for (MeshLib::Element *e : vec_fracutre_elements)
+    {
+        for (unsigned i=0; i<e->getNumberOfBaseNodes(); i++)
+        {
+            MeshLib::Node const* node = e->getNode(i);
+            for (unsigned j=0; j<node->getNumberOfElements(); j++)
+            {
+                // only matrix elemetns
```
Comment (wenqing) [inline ProcessLib/SmallDeformationWithLIE/Common/MeshUtils.cpp]: element

File: Tests/ProcessLib/TestLIE.cpp [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+std::unique_ptr<MeshLib::Line> createY()
+{
+    return createLine({{0.0, -1.0, 0.0}}, {{0.0,  1.0, 0.0}});
+}
+
+std::unique_ptr<MeshLib::Line> createXY()
+{
+    // 45degree inclined
+    return createLine({{0.0, 0.0, 0.0}}, {{2./sqrt(2), 2./sqrt(2), 0.0}});
```
Comment (wenqing) [inline Tests/ProcessLib/TestLIE.cpp]: If it 2./sqrt(2), suggest to use sqrt(2.)

Comment (norihiro-w) [inline Tests/ProcessLib/TestLIE.cpp]: 2./sqrt(2) is used to make the line length 2. This is same in other functions.

File: ProcessLib/SmallDeformationWithLIE/Common/HMatrixUtils.h (hunk 1 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/HMatrixUtils.h]: these declarations are here because they are used in the public part.

File: ProcessLib/SmallDeformationWithLIE/Common/Utils.cpp [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace ProcessLib
+{
+namespace SmallDeformationWithLIE
+{
+
+namespace
+{
+
+void computeNormalVector2D(MeshLib::Element const& e, Eigen::Vector3d & normal_vector)
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/Common/Utils.cpp]: Is the formula for lines only?
Since we faced some problems with the orientation of normal vectors recently: Is the normal vector oriented in the right direction, i.e., is it consistent with our current code? (Maybe have a small discussion about it?)

File: ProcessLib/SmallDeformationWithLIE/Common/Utils.h [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+/// compute physical coordinates from the given shape vector, i.e. from the natural coordinates
+template <typename Derived>
+MathLib::Point3d computePhysicalCoordinates(MeshLib::Element const&e, Eigen::MatrixBase<Derived> const& shape)
+{
+    MathLib::Point3d pt;
+    for (unsigned i=0; i<e.getNumberOfNodes(); i++)
+    {
+        MeshLib::Node const& node = *e.getNode(i);
+        for (unsigned j=0; j<3; j++)
```
Comment (chleh) [inline ProcessLib/SmallDeformationWithLIE/Common/Utils.h]: Minor: `shapeFunctionInterpolate()` could be used for that.

File: Tests/ProcessLib/TestLIE.cpp (hunk 3 above)
Comment (chleh) [inline Tests/ProcessLib/TestLIE.cpp]: but sqrt(2) == 2/sqrt(2). IMHO minor importance.

File: Tests/ProcessLib/TestLIE.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+{
+typedef MeshLib::Line ElementType;
+const unsigned e_nnodes = ElementType::n_all_nodes;
+
+std::unique_ptr<MeshLib::Line> createLine(
+    std::array<double, 3> const& a, std::array<double, 3> const& b)
+{
+    MeshLib::Node** nodes = new MeshLib::Node*[e_nnodes];
+    nodes[0] = new MeshLib::Node(a);
+    nodes[1] = new MeshLib::Node(b);
```
Comment (chleh) [inline Tests/ProcessLib/TestLIE.cpp]: Memory leak?

**Comment by endJunction** [comment]
Please rebase to 6.0.7 before merging.

File: ProcessLib/SmallDeformationWithLIE/Common/FractureProperty.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    /// Initial aperture
+    ProcessLib::Parameter<double>* aperture0 = nullptr;
+};
+
+/// configure fracture property based on a fracture element
+inline void setFractureProperty(unsigned dim, MeshLib::Element const& e,
+                                FractureProperty &frac_prop)
+{
+    for (int j=0; j<3; j++)
+        frac_prop.point_on_fracture[j] = e.getNode(0)->getCoords()[j];
```
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/FractureProperty.h]: As it assumes a fracture forms a straight line, it doesn't matter whether I take the first or second node. I'll add a comment there.

File: ProcessLib/SmallDeformationWithLIE/Common/FractureProperty.h [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+    computeNormalVector(e, frac_prop.normal_vector);
```
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/FractureProperty.h]: not necessary currently

File: ProcessLib/SmallDeformationWithLIE/Common/LevelSetFunction.cpp [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ */
+
+#include "LevelSetFunction.h"
+
+#ifndef Q_MOC_RUN // to avoid Qt4 bug, https://bugreports.qt.io/browse/QTBUG-22829
```
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/LevelSetFunction.cpp]: I'm not really sure when it is really required. maybe @bilke has some information about it.

File: ProcessLib/SmallDeformationWithLIE/Common/LevelSetFunction.cpp [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+} // no named namespace
+
+namespace ProcessLib
+{
+namespace SmallDeformationWithLIE
+{
+
+double calculateLevelSetFunction(
+        FractureProperty const& frac, double const* x_)
```
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/LevelSetFunction.cpp]: it's possible but not better, because the function will be used with MathLib::Point3d.

File: ProcessLib/SmallDeformationWithLIE/Common/LevelSetFunction.h [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+namespace SmallDeformationWithLIE
+{
+
+struct FractureProperty;
+
+/// calculate the level set function
+/// \f$ \psi(\mathbf{x}) = H(|\mathbf{x} - \mathbf{x_d}| \mathrm{sign}[\mathbf{n_d} \cdot (\mathbf{x}-\mathbf{x_d}]) \f$
+/// where \f$H(u)\f$ is the Heaviside step function, \f$\mathbf{x_d}\f$ is a point on the fracture plane,
+/// and \f$\mathbf{n_d}\f$ is the normal vector of a fracture plane
+double calculateLevelSetFunction(
```
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/LevelSetFunction.h]: yes

File: ProcessLib/SmallDeformationWithLIE/Common/Utils.cpp (hunk 4 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/Utils.cpp]: - it's only for lines.
- the normal vector is oriented in the left direction such that there is no rotation required for usual 1D line elements (I mean a result of `computeRotationMatrix2D()`)

File: ProcessLib/SmallDeformationWithLIE/Common/Utils.h (hunk 5 above)
Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/Utils.h]: I wouldn't use it for the two lines.

File: ProcessLib/SmallDeformationWithLIE/Common/FractureProperty.h [hunk 12]
```diff
@@ -0,0 +1,51 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2016, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#pragma once
```
Comment (endJunction) [inline ProcessLib/SmallDeformationWithLIE/Common/FractureProperty.h]: There is https://github.com/ufz/ogs/issues/1376 and so far only two people liked it...

Comment (norihiro-w) [inline ProcessLib/SmallDeformationWithLIE/Common/FractureProperty.h]: I would use the include guard in this PR. Changes to pragma once should be done in another PR.

**Comment by norihiro-w** [comment]
@chleh I can give an overview of the LIE today or tomorrow if we have time. @endJunction please arrange it

**Comment by endJunction** [comment]
I don't think there will be more comments. Replace the pragma once and :fast_forward:
